cpp/include/rapidsmpf/streaming/core/channel.hpp

```diff
@@ -173,7 +173,20 @@ class Channel {
      * @return A coroutine representing the completion of the shutdown drain.
      */
     Node drain(std::unique_ptr<coro::thread_pool>& executor) {
-        return rb_.shutdown_drain(executor);
+        if (rb_.is_shutdown()) {
+            co_return;
+        }
+
+        // Ensure executor progresses until ring buffer is empty, thus ensuring a consumer
+        // that died prematurely (e.g., due to an uncatched exception) does not hang
+        // indefinitely.
+        while (!rb_.empty() && !rb_.is_shutdown()) {
+            co_await executor->yield();
+        }
+
+        if (!rb_.is_shutdown()) {
+            co_await rb_.shutdown();
+        }
```

Comment:
This looks very similar to `shutdown_drain` impl. 
https://github.com/jbaldwin/libcoro/blob/main/include/coro/ring_buffer.hpp#L304-L309

It's not very clear to me why `shutdown_drain` causes a hang and this change resolves it. Maybe add more details?

Reply:
I actually think I have a standalone reproducer now without RapidsMPF. I'll clean it up tomorrow and write an issue in the libcoro summarizing the problem and the potential fix from above.

Reply:
Yeah, I think you're right. I have a simple repro if you want me to submit

Reply:
Opened https://github.com/jbaldwin/libcoro/issues/401 with the summary/repro/suggested solution.